File: src/Leetcode/medium/LongestPalindromicSubstring.java
package Leetcode.medium;

import Test.Test;

import java.util.*;

public class LongestPalindromicSubstring {
    public static void main(String[] args) {
        new LongestPalindromicSubstring().test();
    }

    private void test() {
        String input;
        String expect;
        String result;

        input = "babad";
        expect = "bab";
        result = longestPalindrome(input);
        Test.test(result, expect).printResult();

        input  = "cbbd";
        expect = "bb";
        result = longestPalindrome(input);
        Test.test(result, expect).printResult();

        input  = "a";
        expect ="a";
        result = longestPalindrome(input);
        Test.test(result, expect).printResult();

        input  = "ac";
        expect = "a";
        result = longestPalindrome(input);
        Test.test(result, expect).printResult();

        input  = "aa";
        expect = "aa";
        result = longestPalindrome(input);
        Test.test(result, expect).printResult();

        input  = "xdxtfdaarvvznrptwicdldmrmwbdrmyppvamdvofujthfcmkcugvodmlvubgvodectwzparprifwgwfvddlrfdnrpspirtyvxqvbqiglugbmzoyzcfkvbjdrdqqxxzutebgoacczuhopvzjfrnfsylgfgkbmbjqqyggtdjcvjbvpspkjcezanajjzabfidndfdpkuamwvxrbpxzoivlnkwdxedtfnmvicmzebwktpktokibeycbpqzejddwnvimmbzupyxwmrgdbmcujadfexcchdkfvkxsdwkuwuxzhpnjgmqbmidcwywjgcsbydixyxcclcbrzjvrmlrzgmbviifllouykovscaufvxovwmmgubshtoizbwtcpqzwchtkmkjfneuybfglywfrorhmfdgvjdsmegtoytsivnuaceszpfsxgddbweckgziahkslykgdkztmpapnoyawqtyrdcuzaxcohohapektyfbfhrsdnjbgjvwvqpcikdnlkdogsinkfpymkkdburnbksnqfjgjlacqpfqlhsjhhoccdkrjipqwzsxmpjughaqchzlrqkogkryqkuuxhzchovebzgeekuflcgvxugnxcvugqlstmnljlvxonkybmzjmnsvvwfztcplgikptnppbzeygbmdsyimsntveojwsejmastiovbctdkdlfvpyzihhxishtveflnmamlnzqroxknrrkkfpveyzvvasdznykygrpbfkbinrrvheekeumlvlgalqelspvpiydqkwduckimyhpzsxlcpkbvgwmwnasdxuupdhcmxjoushcvcnjyrmuemuydyywpvzhkxsqszaqhnbhjwsokkpployomoawtr";
        expect = "fwgwf";
        result = longestPalindrome(input);
        Test.test(result, expect).printResult();
    }

    public String longestPalindrome(String s) {
        String longestPalindrome = "";
        for(int i = 0; i < s.length(); i++) {
            String current = s.substring(i);
            String palindrome = getLongestPalindromeString(current);

            if(longestPalindrome.length() < palindrome.length()) {
                longestPalindrome = palindrome;
            }
            if(longestPalindrome.length() >= s.length() - i + 1) {
                break;
            }
        }
        return longestPalindrome;
    }

    private String getLongestPalindromeString(String s) {
        List<Integer> list = getFirstCharIndexList(s);

        for(int index : list) {
            String current = s.substring(0, index+1);
            if(isPalindromic(current)) {
                return current;
            }
        }

        return "";
    }

    private List<Integer> getFirstCharIndexList(String s) {
        List<Integer> list = new ArrayList<>();

        for(int i = 0; i < s.length(); i++) {
            if(s.charAt(0) == s.charAt(i)) {
                list.add(i);
            }
        }
        Collections.reverse(list);
        return list;
    }

    private boolean isPalindromic(String target) {
        for(int i = 0; i <= target.length()/2; i++) {
            if(target.charAt(i) != target.charAt(target.length()-i-1)) {
                return false;
            }
        }
        return true;
    }
}

File: src/Leetcode/medium/LongestPalindromicSubstring.py
import unittest


class Solution:
    def longestPalindrome(self, s: str) -> str:
        odd = []
        even = []

        for i in range(len(s)-1):
            odd.append((i, i))
            if s[i] == s[i+1]:
                even.append((i, i+1))

        result = s[even[0][0]:even[0][1]+1] if even else s[0]

        while odd:
            cur = []
            for (i, j) in odd:
                if i <= 0 or j >= len(s)-1:
                    continue
                if s[i-1] == s[j+1]:
                    cur.append((i-1, j+1))
            if cur:
                result = s[cur[0][0]:cur[0][1]+1]
            odd = cur

        while even:
            cur = []
            for (i, j) in even:
                if i <= 0 or j >= len(s)-1:
                    continue
                if s[i-1] == s[j+1]:
                    cur.append((i-1, j+1))
            if cur and len(result) < len(s[cur[0][0]:cur[0][1]+1]):
                result = s[cur[0][0]:cur[0][1]+1]
            even = cur

        return result


class TestSolution(unittest.TestCase):
    def setUp(self):
        self.solution = Solution()

    def test_case1(self):
        s = "babad"
        expect = "bab"
        result = self.solution.longestPalindrome(s)
        self.assertEqual(expect, result)

    def test_case2(self):
        s = "cbbd"
        expect = "bb"
        result = self.solution.longestPalindrome(s)
        self.assertEqual(expect, result)

    def test_case3(self):
        s = "a"
        expect = "a"
        result = self.solution.longestPalindrome(s)
        self.assertEqual(expect, result)

    def test_case4(self):
        s = "ac"
        expect = "a"
        result = self.solution.longestPalindrome(s)
        self.assertEqual(expect, result)

    def test_case5(self):
        s = "aa"
        expect = "aa"
        result = self.solution.longestPalindrome(s)
        self.assertEqual(expect, result)

    def test_case6(self):
        s = "xdxtfdaarvvznrptwicdldmrmwbdrmyppvamdvofujthfcmkcugvodmlvubgvodectwzparprifwgwfvddlrfdnrpspirtyvxqvbqiglugbmzoyzcfkvbjdrdqqxxzutebgoacczuhopvzjfrnfsylgfgkbmbjqqyggtdjcvjbvpspkjcezanajjzabfidndfdpkuamwvxrbpxzoivlnkwdxedtfnmvicmzebwktpktokibeycbpqzejddwnvimmbzupyxwmrgdbmcujadfexcchdkfvkxsdwkuwuxzhpnjgmqbmidcwywjgcsbydixyxcclcbrzjvrmlrzgmbviifllouykovscaufvxovwmmgubshtoizbwtcpqzwchtkmkjfneuybfglywfrorhmfdgvjdsmegtoytsivnuaceszpfsxgddbweckgziahkslykgdkztmpapnoyawqtyrdcuzaxcohohapektyfbfhrsdnjbgjvwvqpcikdnlkdogsinkfpymkkdburnbksnqfjgjlacqpfqlhsjhhoccdkrjipqwzsxmpjughaqchzlrqkogkryqkuuxhzchovebzgeekuflcgvxugnxcvugqlstmnljlvxonkybmzjmnsvvwfztcplgikptnppbzeygbmdsyimsntveojwsejmastiovbctdkdlfvpyzihhxishtveflnmamlnzqroxknrrkkfpveyzvvasdznykygrpbfkbinrrvheekeumlvlgalqelspvpiydqkwduckimyhpzsxlcpkbvgwmwnasdxuupdhcmxjoushcvcnjyrmuemuydyywpvzhkxsqszaqhnbhjwsokkpployomoawtr";
        expect = "fwgwf"
        result = self.solution.longestPalindrome(s)
        self.assertEqual(expect, result)
